kissmp-server/src/lib.rs

```diff
@@ -634,71 +643,58 @@ pub fn upnp_pf(port: u16) -> Option<u16> {
-                            };
-
-                            if ipv4_addr.ip().is_private() {
-                                valid_ips.push(ipv4_addr);
-                            } else {
-                                continue;
-                            }
-                        }
-                        // v6 Addresses are not compatible with uPnP
+    let mut valid_ips = Vec::new();
+    for interface in ifs {
+        for iface_addr in interface.addresses.iter() {
+            match iface_addr.mask {
+                Some(SocketAddr::V4(ipv4_mask)) => {
+                    let ipv4_addr = match iface_addr.address {
+                        Some(SocketAddr::V4(ipv4_addr)) => ipv4_addr,
                         _ => continue,
+                    };
+
+                    if ipv4_addr.ip().is_private() {
+                        valid_ips.push(ipv4_addr);
+                    } else {
+                        continue;
                     }
                 }
+                // v6 Addresses are not compatible with uPnP
+                _ => continue,
             }
+        }
+    }
 
-            println!(
-                "uPnP: We are going to try the following IPs: {:#?}",
-                valid_ips
-            );
-
-            if valid_ips.is_empty() {
-                eprintln!("uPnP: No interfaces have a valid IP.");
-                return None;
-            }
-
-            for mut ip in valid_ips {
-                ip.set_port(port);
-                println!("uPnP: Trying {}", ip);
-                match gateway.add_port(igd::PortMappingProtocol::UDP, port, SocketAddr::V4(ip), 0, "KissMP Server")
-                {
-                    Ok(()) => return Some(port),
-                    Err(e) => match e {
-                        igd::AddPortError::PortInUse => {
-                            gateway.remove_port(igd::PortMappingProtocol::UDP, port);
-                            gateway.add_port(igd::PortMappingProtocol::UDP, port, SocketAddr::V4(ip), 0, "KissMP Server");
-                            return Some(port)
-                        },
-                        _ => {
-                            eprintln!("uPnP Error: {:?}", e);
-                        }
+    println!(
+        "uPnP: We are going to try the following IPs: {:#?}",
+        valid_ips
+    );
+
+    if valid_ips.is_empty() {
+        Err(UPnPError::NoValidInterfaceIPs)
+    } else {
+        for mut ip in valid_ips {
+            ip.set_port(port);
+            println!("uPnP: Trying {}", ip);
+            loop {
+                match gateway.add_port(
+                    igd::PortMappingProtocol::UDP,
+                    port,
+                    SocketAddr::V4(ip),
+                    0,
+                    "KissMP Server"
+                ) {
+                    Ok(()) => return Ok(port),
+                    Err(igd::AddPortError::PortInUse) => {
+                        gateway.remove_port(igd::PortMappingProtocol::UDP, port)?;
```

Comment:
This could infinite loop from misbehaving gateways if the gateway silently doesn't remove the port. Try limit?